File: mc-java/src/main/java/io/spine/tools/mc/java/gradle/Artifacts.java
/*
 * Copyright 2021, TeamDev. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Redistribution and use in source and/or binary forms, with or without
 * modification, must retain the above copyright notice and the following
 * disclaimer.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package io.spine.tools.mc.java.gradle;

import io.spine.tools.gradle.Artifact;
import io.spine.tools.gradle.Dependency;
import io.spine.tools.gradle.DependencyVersions;
import io.spine.tools.gradle.ThirdPartyDependency;

import static io.spine.tools.gradle.Artifact.SPINE_TOOLS_GROUP;

final class Artifacts {

    private static final String JAR_EXTENSION = "jar";
    private static final String GRPC_GROUP = "io.grpc";
    private static final String GRPC_PLUGIN_NAME = "protoc-gen-grpc-java";
    private static final String MC_JAVA_NAME = "mc-java";
    private static final String EXECUTABLE_CLASSIFIER = "exe";

    static final String SPINE_PROTOC_PLUGIN_NAME = "mc-java-protoc";

    /**
     * The name of the Maven artifact of the Model Compiler Java Checks.
     */
    static final String MC_JAVA_CHECKS_ARTIFACT = "mc-java-checks";

    private static final DependencyVersions versions = DependencyVersions.loadFor("mc-java");

    /**
     * Prevents the utility class instantiation.
     */
    private Artifacts() {
    }

    static Artifact mcJavaChecks() {
        String version = mcJavaVersion();
        return Artifact.newBuilder()
                .useSpineToolsGroup()
                .setName(MC_JAVA_CHECKS_ARTIFACT)
                .setVersion(version)
                .build();
    }

    static Artifact gRpcProtocPlugin() {
        Dependency gRpcPlugin = new ThirdPartyDependency(GRPC_GROUP, GRPC_PLUGIN_NAME);
        return gRpcPlugin.withVersionFrom(versions);
    }

    static Artifact spineProtocPlugin() {
        String version = mcJavaVersion();
        return Artifact.newBuilder()
                .useSpineToolsGroup()
                .setName(SPINE_PROTOC_PLUGIN_NAME)
                .setVersion(version)
                .setClassifier(EXECUTABLE_CLASSIFIER)
                .setExtension(JAR_EXTENSION)
                .build();
    }

    public static String mcJavaVersion() {
        Dependency self = new ThirdPartyDependency(SPINE_TOOLS_GROUP, MC_JAVA_NAME);
        return versions.versionOf(self)
                       .orElseThrow(IllegalStateException::new);
    }
}

File: mc-java/src/main/kotlin/io/spine/tools/mc/java/gradle/Artifacts.kt
/*
 * Copyright 2022, TeamDev. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Redistribution and use in source and/or binary forms, with or without
 * modification, must retain the above copyright notice and the following
 * disclaimer.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

@file:JvmName("Artifacts")

package io.spine.tools.mc.java.gradle

import io.spine.tools.gradle.Artifact
import io.spine.tools.gradle.Artifact.SPINE_TOOLS_GROUP
import io.spine.tools.gradle.Dependency
import io.spine.tools.gradle.DependencyVersions
import io.spine.tools.gradle.ThirdPartyDependency

/**
 * This file defines utilities for generating instances of [Artifact].
 */

private const val JAR_EXTENSION = "jar"
private const val GRPC_GROUP = "io.grpc"
private const val GRPC_PLUGIN_NAME = "protoc-gen-grpc-java"
private const val MC_JAVA_NAME = "spine-mc-java"
private const val ALL_CLASSIFIER = "all"

/**
 * The name of the Maven artifact containing both Spine Protobuf compiler plugin
 * and `modelCompiler` plugin.
 */
internal const val SPINE_MC_JAVA_ALL_PLUGINS_NAME = "spine-mc-java-plugins"

private val versions = DependencyVersions.loadFor(MC_JAVA_NAME)

/**
 * The Maven artifact of the gRPC Protobuf compiler plugin.
 */
@get:JvmName("gRpcProtocPlugin")
internal val gRpcProtocPlugin: Artifact by lazy {
    val gRpcPlugin: Dependency = ThirdPartyDependency(GRPC_GROUP, GRPC_PLUGIN_NAME)
    gRpcPlugin.withVersionFrom(versions)
}

/**
 * The Maven artifact containing the `spine-mc-java-plugins:all` fat JAR artifact.
 */
@get:JvmName("spineJavaAllPlugins")
internal val spineJavaAllPlugins: Artifact by lazy {
    Artifact.newBuilder()
        .useSpineToolsGroup()
        .setName(SPINE_MC_JAVA_ALL_PLUGINS_NAME)
        .setVersion(mcJavaVersion)
        .setClassifier(ALL_CLASSIFIER)
        .setExtension(JAR_EXTENSION)
        .build()
}

private const val VALIDATION_GROUP = "io.spine.validation"

private val validationJavaDependency =
    ThirdPartyDependency(VALIDATION_GROUP, "spine-validation-java")

private val mcJavaProtoDataParamsDependency =
    ThirdPartyDependency(SPINE_TOOLS_GROUP, "spine-mc-java-protodata-params")


private val validationVersion: String by lazy {
    versions.versionOf(validationJavaDependency).orElseThrow()
}

/**
 * The Maven artifact containing the `spine-validation-java` module.
 */
@get:JvmName("validationJava")
internal val validationJava: Artifact by lazy {
    Artifact.newBuilder()
        .setName(validationJavaDependency.name())
        .setGroup(validationJavaDependency.groupId())
        .setVersion(validationVersion)
        .build()
}

/**
 * The Maven artifact containing the `spine-mc-java-protodata-params` module.
 */
@get:JvmName("mcJavaProtoDataParams")
internal val mcJavaProtoDataParams: Artifact by lazy {
    Artifact.newBuilder()
        .setName(mcJavaProtoDataParamsDependency.name())
        .setGroup(mcJavaProtoDataParamsDependency.groupId())
        .setVersion(mcJavaVersion)
        .build()
}

/**
 * The version of the Model Compiler Java modules.
 *
 * This is the version of all the modules declared in this project.
 */
@get:JvmName("mcJavaVersion")
internal val mcJavaVersion: String by lazy {
    val self: Dependency = ThirdPartyDependency(SPINE_TOOLS_GROUP, MC_JAVA_NAME)
    versions.versionOf(self)
        .orElseThrow { IllegalStateException("Unable to load versions of ${self}.") }
}
